src/arch/riscv/irqc/aia/vaplic.c

```diff
@@ -0,0 +1,1057 @@
+        } else {
+            vaplic_update_hart_line(vcpu);
+        }
+    }
+    spin_unlock(&virqc->lock);
+}
+
+/**
+ * @brief Clear the pending bits for interrupts [32*reg:(32*reg)+31]
+ * 
+ * @param vcpu virtual cpu 
+ * @param reg  regiter index
+ * @param new_val value with intp to be cleared per bit
+ */
+static void vaplic_set_in_clrip(struct vcpu *vcpu, uint8_t reg, uint32_t new_val){
+    struct virqc *virqc = &vcpu->vm->arch.virqc;
+    spin_lock(&virqc->lock);
+    if (reg < APLIC_NUM_CLRIx_REGS && 
+        vaplic_get_setip(vcpu, reg) != new_val) {
+        if (reg == 0) new_val &= 0xFFFFFFFE;
+        virqc->setip[reg] &= ~new_val;
+        for(int i = 0; i < APLIC_MAX_INTERRUPTS/APLIC_NUM_CLRIx_REGS; i++){
+            if(vaplic_get_hw(vcpu,i)){
+                if(!get_bit_from_reg(virqc->setip[reg], i) && ((new_val >> i) & 1)){
+                    aplic_set_clripnum(i);
+                }
+            } else {
+                vaplic_update_hart_line(vcpu);
+            }
+        }
+    }
+    spin_unlock(&virqc->lock);
+}
+
+/**
+ * @brief Get the rectified input values per source
+ *        NOTE: Not implemented as stated! 
+ *        TODO: return the pending bits?
+ * 
+ * @param vcpu virtual cpu 
+ * @param reg regiter index
+ * @return uint32_t value with rectified intp per bit
+ */
+static uint32_t vaplic_get_in_clrip(struct vcpu *vcpu, uint8_t reg){
+    uint32_t ret = 0;
+    struct virqc * virqc = &vcpu->vm->arch.virqc;
+    if (reg < APLIC_NUM_CLRIx_REGS) ret = virqc->in_clrip[reg];
+    return ret;
+}
+
+/**
+ * @brief Clear the pending bits for a given interrupt
+ * 
+ * @param vcpu virtual cpu
+ * @param new_val value w/ the interrupt source number
+ */
+static void vaplic_set_clripnum(struct vcpu *vcpu, uint32_t new_val){
+    struct virqc *virqc = &vcpu->vm->arch.virqc;
+    spin_lock(&virqc->lock);
+    if (new_val != 0 && new_val < APLIC_MAX_INTERRUPTS && 
+        get_bit_from_reg(virqc->setip[new_val/32], new_val)) {
+        clr_bit_from_reg(&virqc->setip[new_val/32], new_val%32);
+        if(vaplic_get_hw(vcpu,new_val)){
+            aplic_set_clripnum(new_val);
+        } else {
+            vaplic_update_hart_line(vcpu);
+        }
+    }
+    spin_unlock(&virqc->lock);
+}
+
+/**
+ * @brief Get the enabled bits for interrupts [32*reg:(32*reg)+31]
+ * 
+ * @param vcpu virtual cpu
+ * @param reg regiter index
+ * @return uint32_t value with enabled bit mapped per bit
+ */
+static uint32_t vaplic_get_setie(struct vcpu *vcpu, uint32_t reg){
+    uint32_t ret = 0;
+    struct virqc * virqc = &vcpu->vm->arch.virqc;
+    if (reg < APLIC_NUM_SETIx_REGS) ret = virqc->setie[reg];
+    return ret;
+}
+
+/**
+ * @brief Set the enabled bits for interrupts [32*reg:(32*reg)+31]
+ * 
+ * @param vcpu virtual cpu
+ * @param reg regiter index
+ * @param new_val value with enbaled bit mapped per bit
+ */
+static void vaplic_set_setie(struct vcpu *vcpu, uint8_t reg, uint32_t new_val){
```

Comment:
This code is current as shown below:

```
static void vaplic_set_setie(struct vcpu *vcpu, uint8_t reg, uint32_t new_val){
    struct vaplic *vaplic = &vcpu->vm->arch.vaplic;

    spin_lock(&vaplic->lock);
    if (reg == 0) new_val &= MASK_INTP_ZERO;
    if (reg < APLIC_NUM_SETIx_REGS && 
        vaplic_get_setie(vcpu, reg) != new_val) {
        for(size_t i = (reg*APLIC_NUM_INTP_PER_REG); 
            i < (reg*APLIC_NUM_INTP_PER_REG) + APLIC_NUM_INTP_PER_REG; i++){
            if (vaplic_get_active(vcpu, i)){
                if(vaplic_get_hw(vcpu,i)){
                    aplic_set_enbl(i);
                    if(aplic_get_enbl(i)){
                        BIT32_SET_INTP(vaplic->ie, i);
                    }
                } else {
                    BIT32_SET_INTP(vaplic->ie, i);
                }
            }
            vaplic_update_hart_line(vcpu, GET_HART_INDEX(vcpu, i));
        }
    }
    spin_unlock(&vaplic->lock);
}
```

The thing here is the same as in `vaplic_set_setip`. 
Another thing is: in the code above, when the interrupt is connected to hw, I only update the virt if the hw was succ. written. However, I believe we don't need to do this check for the enable array. So we could just mask the new_val with the active array, write the new_val in vaplic->ie, mask it again with the hw array and write it in aplic. Do you think it is a good approach?

In short: 

```
        new_val &= vaplic->active[reg];
        vaplic->ie[reg] = new_val;
        new_val &= vaplic->hw[reg];
        aplic_set32_enbl(reg, new_val);
        vaplic_update_hart_line(vcpu, UPDATE_ALL_HARTS);
```